README.md
## Rack::Clicky ##

Rack Middleware to embed the Clicky (http://www.getclicky.com) tracking code.


### Installation.

    gem install rack_clicky

or if you're using Bundler:

    gem "rack_clicky", require: 'rack/clicky'
    
### Usage ###

    require 'rack/clicky'

    use Rack::Clicky, tracker: '000000'
    app = lambda { |env|
      output = '<html><body><p>Awesome Body</p></body></html>'
      [200, { 'Content-Type' => 'text/html' }, output] 
    }
    run app

There are more examples in the `examples` directory.

### Note on Patches/Pull Requests ###
 
* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

### Copyright and Licence ###

Copyright (c) 2010 Mark Turner. See LICENSE for details.


CHANGES.md
# CHANGES #

## v1.0.6 - Tuesday the 30th of January 2018 ##

- Rack removed the bytesize method in favour of String's,
  this change takes care of both possibilities so will run on
  either version.

----

test/helper.rb
require 'rubygems'
require 'minitest/autorun'
require 'minitest/pride'
require 'rack/mock'

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'rack/clicky'


spec/spec_helper.rb
# encoding: UTF-8

require 'rspec'
Spec_dir = File.expand_path( File.dirname __FILE__ )

unless Kernel.respond_to?(:require_relative)
  module Kernel
    def require_relative(path)
      require File.join(File.dirname(caller[0]), path.to_str)
    end
  end
end

# code coverage
require 'simplecov'
SimpleCov.start do
  add_filter "/vendor/"
  add_filter "/bin/"
  add_filter "/examples/"
end

require "rack/test"
ENV['RACK_ENV'] ||= 'test'
ENV["EXPECT_WITH"] ||= "racktest"



require "logger"
logger = Logger.new STDOUT
logger.level = Logger::DEBUG
logger.datetime_format = '%a %d-%m-%Y %H%M '
LOgger = logger


Dir[ File.join( Spec_dir, "/support/**/*.rb")].each do |f| 
  logger.info "requiring #{f}"
  require f
end


RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
end


spec/rack_clicky_spec.rb
# encoding: UTF-8

require 'spec_helper'
require_relative "../lib/rack/clicky.rb"

describe Rack::Clicky do
  describe "Embedding clicky" do
    include_context "Application"
    let(:link) { <<STR
<a title="Real Time Web Analytics" href="http://clicky.com/000000"><img alt="Real Time Web Analytics" src="//static.getclicky.com/media/links/badge.gif" border="0" /></a>
STR
    }
    let(:async_script) { s = <<STR
<script type=\"text/javascript\">
  var clicky_site_ids = clicky_site_ids || [];
  clicky_site_ids.push(000000);
  (function() {
    var s = document.createElement('script');
    s.type = 'text/javascript';
    s.async = true;
    s.src = '//static.getclicky.com/js';
    ( document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0] ).appendChild( s );
  })();
</script>
<noscript><p><img alt=\"Clicky\" width=\"1\" height=\"1\" src=\"//in.getclicky.com/000000ns.gif\" /></p></noscript>

</body>
STR
      s.strip!
    }
    let(:sync_script) { s= <<STR
<script src="//static.getclicky.com/js" type="text/javascript"></script>
<script type="text/javascript">try{ clicky.init(000000); }catch(e){}</script>
<noscript><p><img alt="Clicky" width="1" height="1" src="//in.getclicky.com/000000ns.gif" /></p></noscript>

</body>
STR
      s.strip!
    }

    describe "Synchronous" do
      before :all do
        Example.app.class.clear_caches
      end
      context "Given a false regarding async" do
        include_context "Synchronous"
        let(:script) { "#{link}#{sync_script}" }
        context "a 200 status and html served" do
          before{ get "/html", {},{"HTTP_ACCEPT" => "text/html" } }
          it_should_behave_like "Any route"
          subject { last_response.body }
          it { should include script }
          describe "async setting" do
            subject { app.class.async }      
            it { should be_false }
          end
        end  
        context "a 200 status and xhtml served" do
          before{ get "/xhtml", {},{"HTTP_ACCEPT" => "application/xhtml+xml" } }
          it_should_behave_like "Any route"
          subject { last_response.body }
          it { should include script }
          describe "async setting" do
            subject { app.class.async }      
            it { should be_false }
          end
        end 
        context "a 200 status and xml served" do
          before{ get "/xml", {},{"HTTP_ACCEPT" => "application/xml" } }
          it_should_behave_like "Any route"
          subject { last_response.body }
          it { should_not include script }
          describe "async setting" do
            subject { app.class.async }      
            it { should be_false }
          end
        end
  
      end
    end

    describe "Asynchronous" do
      before :all do
        Example.app.class.clear_caches
      end
      context "Given a true regarding async" do
        include_context "Asynchronous"
        let(:script) { "#{link}#{async_script}" }
        context "a 200 status and html served" do
          describe "async setting" do
            subject { app.class.async }      
            it { should be_true }
          end
          before{ get "/html", {},{"HTTP_ACCEPT" => "text/html" } }
          it_should_behave_like "Any route"
          subject { last_response.body }
          it { should include script }
        end  
        context "a 200 status and xhtml served" do
          describe "async setting" do
            subject { app.class.async }      
            it { should be_true }
          end
          before{ get "/xhtml", {},{"HTTP_ACCEPT" => "application/xhtml+xml" } }
          it_should_behave_like "Any route"
          subject { last_response.body }
          it { should include script }
        end 
        context "a 200 status and xml served" do
          describe "async setting" do
            subject { app.class.async }      
            it { should be_true }
          end
          before{ get "/xml", {},{"HTTP_ACCEPT" => "application/xml" } }
          it_should_behave_like "Any route"
          subject { last_response.body }
          it { should_not include script }
        end
  
      end
    end

    describe "Without link" do
      before :all do
        Example.app.class.clear_caches
      end
      include_context "Without link"
      let(:script) { async_script }
      context "a 200 status and html served" do
        before{ get "/html", {},{"HTTP_ACCEPT" => "text/html" } }
        it_should_behave_like "Any route"
        subject { last_response.body }
        it { should include script }
        it { should_not include link }
      end  
    end
  end

end


lib/rack_clicky.rb
require "rack_clicky/version"

# @see http://rack.rubyforge.org/doc/SPEC.html
module Rack

  # A Rack class for helping with the Clicky.com javascript and tracker code. The documentation for {Clicky#initialize} is the important bit.
  class Clicky

    # Produces the script.
    # This also caches the script, as unless the tracking code
    # and/or the async settings change, the script will remain
    # the same.
    def self.script
      @script ||= ( @async ? 
                    ASYNC_SCRIPT : 
                    SYNC_SCRIPT 
                  ).gsub!(/\{\{CODE\}\}/, @tracker)
    end


    # The script tags for the synchronous clicky script.
    SYNC_SCRIPT = <<-EOTC 
    <script src="//static.getclicky.com/js" type="text/javascript"></script>
    <script type="text/javascript">try{ clicky.init({{CODE}}); }catch(e){}</script>
    <noscript><p><img alt="Clicky" width="1" height="1" src="//in.getclicky.com/{{CODE}}ns.gif" /></p></noscript>
    EOTC

    # The script tags for the asynchronous clicky script.
    ASYNC_SCRIPT = <<-STR
      <script type="text/javascript">
      var clicky_site_ids = clicky_site_ids || [];
      clicky_site_ids.push({{CODE}});
      (function() {
        var s = document.createElement('script');
        s.type = 'text/javascript';
        s.async = true;
        s.src = '//static.getclicky.com/js';
        ( document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0] ).appendChild( s );
      })();
      </script>
      <noscript><p><img alt="Clicky" width="1" height="1" src="//in.getclicky.com/{{CODE}}ns.gif" /></p></noscript>
    STR

    # @param [String] tracker The tracking code.
    def self.tracker=( tracker )
      @tracker = tracker
    end

    # @return [String] The tracking code.
    def self.tracker
      @tracker
    end

    # @param [true,false] async True to use the asynchronous script, false for the synchronous script via CDN. Defaults to true (via the constructor).
    def self.async=( async )
      @async = async
    end

    # @return [true,false]
    def self.async
      @async
    end


    # Clears all the cached variables, tracker, async, and script.
    def self.clear_caches
      @async = nil
      @tracker = nil
      @script = nil
    end


    # Use the middleware, just pass in a tracker code.
    # The values given as options are cached.
    # @param [Proc] app
    # @param [Hash] options
    # @option [String] :tracker The tracking code. The app will fail if this is not supplied.
    # @option [true,false] :async Whether to use the asynchronous (default) or the synchronous script.
    # @example
    #   use Rack::Clicky, tracker: "000000" # async is the default
    #
    #   # for synchronous script
    #   use Rack::Clicky, tracker: "000000", async: false
    def initialize( app, options={} )
      fail ArgumentError, "Tracker must be set!" if options[:tracker].nil? || options[:tracker].empty?
      @app, @options  = app, options
      self.class.tracker  ||= @options[:tracker]
      self.class.async = @options.fetch(:async, true)
    end


    # @param [#call] env
    # @return [Array]
    def call( env )
      dup._call(env)
    end


    # Duplicated to make thread safe.
    # @see #call
    def _call( env )
      status, headers, response = @app.call(env)
      
      if should_inject_clicky?( status, headers )
        response = inject_script(response)
        headers['Content-Length'] = calc_content_length(headers, response).to_s if headers["Content-Length"]
      end

      [status, headers, response]
    end


    private


    # @param [Integer] status
    # @param [Hash] headers
    # @return [true,false]
    def should_inject_clicky?( status, headers )
      [200,201].any?{|sc| status == sc }  &&
      headers["Content-Type"]             &&
      ["text/html","application/xhtml"].any? {|ct| 
        headers["Content-Type"].include? ct 
      }
    end


    # Calculates the (new) content length.
    # @param [Hash] headers
    # @param [Rack::Response] response
    # @return [Integer]
    def calc_content_length( headers, response )
      length = response.to_ary.inject(0) { |len, part| 
        len + Rack::Utils.bytesize(part) 
      }
    end


    # Injects the script into the response.
    # @param [Rack::Response] response
    # @return [Array]
    def inject_script( response, body="" )
      response.each { |s| body << s.to_s }
      [body.gsub(/<\/body>/, "#{self.class.script}\n</body>")]
    end

  end
end


lib/rack/clicky.rb
# @see http://rack.rubyforge.org/doc/SPEC.html
module Rack

  # A Rack class for helping with the Clicky.com javascript and tracker code. The documentation for {Clicky#initialize} is the important bit.
  class Clicky

    # Produces the script.
    # This also caches the script, as unless the tracking code
    # and/or the async/link settings change, the script will remain
    # the same.
    def self.script
      @script ||=
        (
          (@with_link ? 
            CLICKY_LINK :
            ""
          ) +
          ( @async ? 
            ASYNC_SCRIPT : 
            SYNC_SCRIPT 
          )
        ).gsub!(/\{\{CODE\}\}/, @tracker)
    end


    CLICKY_LINK = <<STR
<a title="Real Time Web Analytics" href="http://clicky.com/{{CODE}}"><img alt="Real Time Web Analytics" src="//static.getclicky.com/media/links/badge.gif" border="0" /></a>
STR

    # The script tags for the synchronous clicky script.
    SYNC_SCRIPT = <<EOTC 
<script src="//static.getclicky.com/js" type="text/javascript"></script>
<script type="text/javascript">try{ clicky.init({{CODE}}); }catch(e){}</script>
<noscript><p><img alt="Clicky" width="1" height="1" src="//in.getclicky.com/{{CODE}}ns.gif" /></p></noscript>
EOTC

    # The script tags for the asynchronous clicky script.
    ASYNC_SCRIPT = <<STR
<script type="text/javascript">
  var clicky_site_ids = clicky_site_ids || [];
  clicky_site_ids.push({{CODE}});
  (function() {
    var s = document.createElement('script');
    s.type = 'text/javascript';
    s.async = true;
    s.src = '//static.getclicky.com/js';
    ( document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0] ).appendChild( s );
  })();
</script>
<noscript><p><img alt="Clicky" width="1" height="1" src="//in.getclicky.com/{{CODE}}ns.gif" /></p></noscript>
STR


    # @param [String] tracker The tracking code.
    def self.tracker=( tracker )
      @tracker = tracker
    end

    # @return [String] The tracking code.
    def self.tracker
      @tracker
    end


    # @param [true,false] async True to use the asynchronous script, false for the synchronous script via CDN. Defaults to true (via the app's constructor).
    def self.async=( async )
      @async = async
    end

    # @return [true,false]
    def self.async
      @async
    end


    # @param [true,false] async True to show the link to clicky.com. Defaults to true (via the app's constructor).
    def self.with_link=( link )
      @with_link = link
    end

    def self.with_link?
      @with_link
    end


    # Clears all the cached variables, tracker, async, and script.
    def self.clear_caches
      @async = nil
      @tracker = nil
      @script = nil
    end


    # Use the middleware, just pass in a tracker code.
    # The values given as options are cached.
    # @param [Proc] app
    # @param [Hash] options
    # @option options [String] :tracker The tracking code. The app will fail if this is not supplied.
    # @option options [true,false] :async Whether to use the asynchronous (default) or the synchronous script.
    # @option options [true,false] :link Whether to add the link to clicky.com or not. Default is true.
    # @example
    #   use Rack::Clicky, tracker: "000000"
    #   async with the link is the default
    #
    #   # for synchronous script without link
    #   use Rack::Clicky, tracker: "000000", async: false, link: false
    def initialize( app, options={} )
      fail ArgumentError, "Tracker must be set!" if options[:tracker].nil? || options[:tracker].empty?
      @app, @options  = app, options
      self.class.tracker  ||= @options[:tracker]
      self.class.async = @options.fetch(:async, true)
      self.class.with_link = @options.fetch(:link, true)
    end


    # @param [#call] env
    # @return [Array]
    def call( env )
      dup._call(env)
    end


    # Duplicated to make thread safe.
    # @see #call
    def _call( env )
      status, headers, response = @app.call(env)
      
      if should_inject_clicky?( status, headers )
        response = inject_script(response)
        headers['Content-Length'] = calc_content_length(headers, response).to_s if headers["Content-Length"]
      end

      [status, headers, response]
    end


    private


    # If it's xHTML or HTML, then inject the scripts.
    # Anything else, it ignores it.
    # @param [Integer] status
    # @param [Hash] headers
    # @return [true,false]
    def should_inject_clicky?( status, headers )
      [200,201].any?{|sc| status == sc }  &&
      headers["Content-Type"]             &&
      ["text/html","application/xhtml"].any? {|ct| 
        headers["Content-Type"].include? ct 
      }
    end


    # Calculates the (new) content length.
    # @param [Hash] headers
    # @param [Rack::Response] response
    # @return [Integer]
    def calc_content_length( headers, response )
      length = response.to_ary.inject(0) { |len, part|
        if part.respond_to? :bytesize
          len + part.bytesize
        elsif Rack::Utils.respond_to? :bytesize
          len + Rack::Utils.bytesize()
        else
          fail "There has to be an available `bytesize` method for this to work!"
        end
      }
    end


    # Injects the script into the response.
    # @param [Rack::Response] response
    # @return [Array]
    def inject_script( response, body="" )
      response.each { |s| body << s.to_s }
      [ body.sub( %r{</body>}, "#{self.class.script}\n</body>") ]
    end

  end
end
